### PR TITLE
Unit-2.3-JS: Minimizing/Expanding Functionality

### PR DESCRIPTION
**Pull request by nathanjophillips**
## The Purpose
- Gain a basic understanding of javascript
- Apply simple vanilla javascript to project to give if more functionality
- Use javascript to change the layout of the project in a meaningful, UX-improving way

## What It Does

- This feature adds an OpenClose() function to allow the content to be toggled between classes `expand` and `collapse`(the collapse class being styled as `display: hidden;`) giving it a way of expanding and collapsing certain content in the user profile card.

- OpenClose() also changes the button text every time it's clicked by using the `.innerHTML` method.

### 1. Expanding:
- [x] Clicking a button will expand the user profile
- [x] The button is styled in a way that says "hey, I'll expand the profile if you click me"

### 2. Minimizing:
- [x] Clicking a button will minimize the user profile
- [x] The button is styled in a way that says "hey, le'me make this profile a little smaller"

### 3. Expand/Minimize:
- [x]  If minimized, clicking the button will expand the user profile
- [x] If expanded, clicking the button will minimize the user profile
- [x] Have expanding and minimizing functionality run by the same button

### The changes to look for are:
1. After `Expand` button clicked <b>Phone and Email</b> are added below in their own rows
2. After `Collapse` button clicked <b>only ID and Name</b> appear in their own rows
3. Button text shows `Collapse` when all four rows are showing <b>Not</b> when only two rows are showing
4. Button text shows `Expand` when top two rows are showing <b>Not</b> when all four rows are showing

# How To Test
- Clone Repo [nathan-phillips-ux-iop](https://github.com/phillnab/nathan-phillips-ux-iop)
- Checkout branch `nathan-phillips-ux-iop/unit-2.3-js`
- Open <b>only</b> `user-profile.html` and `style.css` to view locally
- Compare your expanded and collapsed views to the following screenshots

## Expanded

1. Showing ID, Name, Phone, and Email data rows
2. Button text displays `Collapse`

screen shot 2019-02-28 at 11 54 02 am

## Collapsed

1. Only showing ID and Name data rows
2. Button text displays `Expand`

screen shot 2019-02-28 at 10 44 06 am

## Checklist

Under penalty of public shaming, I avow that I:

- [x] Reviewed the diff to look for typos, whitespace errors, and console.log() statements
- [ ] Verified that there are no compiler or linter errors or warnings
- [ ] Verified the build in production(optimized) mode
- [x] Updated the docs, if necessary
- [x] Included the appropriate labels
- [x] Referenced applicable issues ()

Browsers tested:
- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge
- [ ] Internet Explorer

### CODE REVIEW

File: html/user-profile.html
Comment (Jropp) [inline html/user-profile.html]: The name `openClose` might not hold up well as you add more code. There might be other items that open and close, like a dropdown menu.

File: html/user-profile.html
Comment (Jropp) [inline html/user-profile.html]: If you are only going to be collapsing/expanding a single card on this page, then you could just use `document.getElementById/getElementByClassName` rather than `document.getElementsById/getElementsByClassname`.

If you are intending to get multiple elements by classname, you are probably going to run into some problems that I'll mention in the next comment.

File: html/user-profile.html
Comment (Jropp) [inline html/user-profile.html]: If you have multiple profile cards this will perform an action on any of the buttons that have the expand class on them at the moment. So a click on a single profile card would have an effect on multiple cards.

And again if you intend to only have a single profile card, you might as well just target that single profile card by id.

My guess here is that a more practical/simpler solution might involve using [event.target](https://developer.mozilla.org/en-US/docs/Web/API/Event/target)

File: html/user-profile.html
Comment (Jropp) [inline html/user-profile.html]: This would be a good spot to try out a [ternary operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Conditional_Operator)

**Review by Jropp — CHANGES_REQUESTED** [review]
The functionality works great, but I think the method and function naming will run you into trouble pretty quick

File: html/user-profile.html
Comment (nathanjophillips) [reply html/user-profile.html]: would a better name be cardCollapse() or cardExpand()? I was trying to keep it somewhat short, but I see what you are saying on how I'll run into future problems

File: html/user-profile.html
Comment (nathanjophillips) [reply html/user-profile.html]: Yeah the intent is to access to just the one element at a time at the moment, but I'd like to figure out how to be able to make it work separately for each card in the list.

File: html/user-profile.html
Comment (nathanjophillips) [reply html/user-profile.html]: I'm a little confused on how targeting events works at the moment. I'll do some more reading on this and see if I can make it work since it'll benefit me in the future.

File: html/user-profile.html
Comment (nathanjophillips) [reply html/user-profile.html]: I'll mess around with this too and see if I can make it work. I'll let you know if I have questions.

File: html/user-profile.html
Comment (kddrake) [reply html/user-profile.html]: @phillnab your 1st priority with naming should be to be as succinct as possible, not necessarily short.  both `cardCollapse()` and `cardExpand()` would be much better choices, assuming that's all the function does. You don't want a function called `cardCollapse` that does _both_, though, so make sure the function does what the function name implies it does.

File: html/user-profile.html
Comment (kddrake) [inline html/user-profile.html]:
```suggestion
          <button id="cardToggleButton" onclick="openClose()">Collapse</button>
```

**Review by kddrake — COMMENTED** [review]
other than my one nit, and all Jason's comments, this looks good by me.

**Comment by nathanjophillips** [comment]
Since this has been taking up more of my time than I'd like it to at the moment,  I'm going go back to what I had for now.  Especially since it's hard to test the targeted event by using only one card, not the list of cards that will be all on one page soon.  So, my plan is to come back to this in the near future so I can continue to make progress in other areas of the project. @kddrake @Jropp  feel free to merge this in when you approve or if you have anything else please let me know.

File: html/user-profile.html
Comment (Jropp) [inline html/user-profile.html]: You could cut this down to a single line using a ternary operator https://www.thoughtco.com/javascript-by-example-use-of-the-ternary-operator-2037394

File: html/user-profile.html
Comment (Jropp) [inline html/user-profile.html]: I hadn't messed with the `classList` property, so new learning for me today!

This adds or removes the collapse class, but the expand class is still in place. Could you also toggle the expand class here?

File: html/user-profile.html
Comment (Jropp) [inline html/user-profile.html:69]: Specificity! I mean there may be some other selector that might let you jump to the parent element you want by targeting an id, but I can't think of/find it at the moment. I'll leave it to Kean on this one.

File: html/user-profile.html
Comment (Jropp) [reply html/user-profile.html]: You could also give the elements an id of collapsable and target that with your query selector on line 69 and then toggle the collapse class. I just don't like the idea of having expanded and collapsed as classes at the same time.

File: html/user-profile.html
Comment (Jropp) [inline html/user-profile.html:20]: Just a couple small tweaks, otherwise looks good! Thanks for hanging in there on my requested changes.

File: html/user-profile.html
Comment (kddrake) [reply html/user-profile.html:69]: I was thinking _maybe_ `document` or `window`? But that's a total stab in the dark.

File: html/user-profile.html
Comment (kddrake) [reply html/user-profile.html]: yes ^^

File: html/user-profile.html
Comment (nathanjophillips) [reply html/user-profile.html:69]: I tried doing that before, but by using `.childen()` it would make an array rather than let me grab the specific element i needed.

Comment (nathanjophillips) [reply html/user-profile.html:69]: and `element` acts as the `event` that's being targeted in this can. Should i change the name to `event`

File: html/user-profile.html
Comment (nathanjophillips) [reply html/user-profile.html]: the problem with the id is that it only allowed me to grab one element at a time rather than hiding both row. and is there a reason toggling the expand class would be better?

File: html/user-profile.html
Comment (kddrake) [reply html/user-profile.html:69]: @phillnab nah, I'm not realizing how we deal with events in polymer isn't as close to vanilla as I had thought. what you have right here is right.

**Review by kddrake — APPROVED** [review]

File: html/user-profile.html
Comment (Jropp) [reply html/user-profile.html]: @phillnab I just don't like the idea of having an expanded and collapsed class at the same time. 

If I were stumbling upon this I would think that the expanded state represents something when it is expanded. As you currently have it it could still have the expand class when it's actually collapsed.

Even setting the class to something like collapsable-row is better imo because then someone coming along later to make changes won't confuse that with styles affecting a state.

**Pull request merged**
